File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/jdbc/StatementSetterGenerator.java
package ru.tinkoff.kora.database.annotation.processor.jdbc;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.FieldFactory;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;

import javax.annotation.Nullable;
import javax.lang.model.element.ExecutableElement;
import java.util.List;

import static ru.tinkoff.kora.annotation.processor.common.CommonUtils.isNullable;

public class StatementSetterGenerator {

    public static CodeBlock generate(ExecutableElement method, QueryWithParameters sqlWithParameters, List<QueryParameter> parameters, @Nullable QueryParameter batchParam, FieldFactory parameterMappers) {
        var b = CodeBlock.builder();
        if (batchParam != null) {
            // one of Iterable<T>, Iterator<T>, Stream<T>
            b.add("for (var _i : $L) {$>\n", batchParam.variable());
        }
        for (int i = 0; i < parameters.size(); i++) {
            var parameter = parameters.get(i);
            if (parameter instanceof QueryParameter.ConnectionParameter) {
                continue;
            }
            var _i = i;
            var parameterName = parameter.name();
            if (parameter instanceof QueryParameter.BatchParameter batchParameter) {
                parameterName = "_i";
                parameter = batchParameter.parameter();
            }
            if (parameter instanceof QueryParameter.SimpleParameter nativeParameter) {
                var sqlParameter = sqlWithParameters.parameters()
                    .stream()
                    .filter(p -> p.methodIndex() == _i)
                    .findFirst()
                    .get();
                var mappingData = CommonUtils.parseMapping(parameter.variable());
                var mapping = mappingData.getMapping(JdbcTypes.PARAMETER_COLUMN_MAPPER);
                var nativeType = JdbcNativeTypes.findNativeType(TypeName.get(parameter.type()));
                if (mapping == null && nativeType != null) {
                    if (isNullable(parameter.variable())) {
                        b.add("if ($L != null) {$>", parameterName);
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.add("\n").add(nativeType.bind("_stmt", parameterName, idx + 1)).add(";");
                        }
                        b.add("$<\n} else {$>");
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.add("\n").add(nativeType.bindNull("_stmt", idx + 1)).add(";");
                        }
                        b.add("$<\n}\n");
                    } else {
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.add(nativeType.bind("_stmt", parameterName, idx + 1)).add(";\n");
                        }
                    }
                } else if (mapping != null && mapping.mapperClass() != null) {
                    var mapper = parameterMappers.get(JdbcTypes.PARAMETER_COLUMN_MAPPER, mapping, parameter.type());
                    for (var idx : sqlParameter.sqlIndexes()) {
                        b.add("$L.set(_stmt, $L, $L);\n", mapper, idx + 1, parameterName);
                    }
                } else {
                    var mapper = parameterMappers.get(JdbcTypes.PARAMETER_COLUMN_MAPPER, parameter.type(), parameter.variable());
                    for (var idx : sqlParameter.sqlIndexes()) {
                        b.add("$L.set(_stmt, $L, $L);\n", mapper, idx + 1, parameterName);
                    }
                }
            }
            if (parameter instanceof QueryParameter.EntityParameter entityParameter) {
                for (var field : entityParameter.entity().columns()) {
                    var sqlParameter = sqlWithParameters.find(field.queryParameterName(entityParameter.name()));
                    if (sqlParameter == null || sqlParameter.sqlIndexes().isEmpty()) {
                        continue;
                    }
                    var accessor = CodeBlock.of("$N.$N()", parameterName, field.accessor()).toString();
                    var mappingData = CommonUtils.parseMapping(field.element());
                    var mapping = mappingData.getMapping(JdbcTypes.PARAMETER_COLUMN_MAPPER);
                    var nativeType = JdbcNativeTypes.findNativeType(TypeName.get(field.type()));
                    if (mapping == null && nativeType != null) {
                        if (isNullable(field.element())) {
                            b.add("if ($L != null) {$>", accessor);
                            for (var idx : sqlParameter.sqlIndexes()) {
                                b.add("\n").add(nativeType.bind("_stmt", accessor, idx + 1)).add(";");
                            }
                            b.add("$<\n} else {$>");
                            for (var idx : sqlParameter.sqlIndexes()) {
                                b.add("\n").add(nativeType.bindNull("_stmt", idx + 1)).add(";");
                            }
                            b.add("$<\n}\n");
                        } else {
                            for (var idx : sqlParameter.sqlIndexes()) {
                                b.add(nativeType.bind("_stmt", accessor, idx + 1)).add(";\n");
                            }
                        }
                    } else if (mapping == null) {
                        var mapper = parameterMappers.get(JdbcTypes.PARAMETER_COLUMN_MAPPER, field.type(), field.element());
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.add("$L.set(_stmt, $L, $L);\n", mapper, idx + 1, accessor);
                        }
                    } else {
                        var mapper = parameterMappers.get(JdbcTypes.PARAMETER_COLUMN_MAPPER, mapping, field.type());
                        for (var idx : sqlParameter.sqlIndexes()) {
                            b.add("$L.set(_stmt, $L, $L);\n", mapper, idx + 1, accessor);
                        }
                    }
                }
            }
        }
        if (batchParam != null) {
            b.add("_stmt.addBatch();$<\n};\n");
        }
        return b.build();
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/jdbc/StatementSetterGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.jdbc

import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.parameterMapperName
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.parseMappingData

object StatementSetterGenerator {

    fun generate(b: FunSpec.Builder, function: KSFunctionDeclaration, queryWithParameters: QueryWithParameters, parameters: List<QueryParameter>, batchParam: QueryParameter?) {
        if (batchParam != null) {
            b.beginControlFlow("for (_batch_%L in %N)", batchParam.name, batchParam.name)
        }
        parameters.forEachIndexed { i, p ->
            var parameter = p
            if (parameter is QueryParameter.ConnectionParameter) {
                return@forEachIndexed
            }
            var parameterName = parameter.name
            if (parameter is QueryParameter.BatchParameter) {
                parameter = parameter.parameter
                parameterName = "_batch_${parameter.name}"
            }
            if (parameter is QueryParameter.ParameterWithMapper) {
                val sqlParameter = queryWithParameters.find(i)!!
                for (idx in sqlParameter.sqlIndexes) {
                    b.addStatement("%N.set(_stmt, %L, %N)", parameterMapperName(function, parameter.variable), idx + 1, parameterName)
                }
                return@forEachIndexed
            }
            if (parameter is QueryParameter.SimpleParameter) {
                val sqlParameter = queryWithParameters.find(i)!!
                val mapping = parameter.variable.parseMappingData().getMapping(JdbcTypes.jdbcParameterColumnMapper)
                val nativeType = JdbcNativeTypes.findNativeType(parameter.type.toTypeName())
                if (nativeType != null && mapping == null) {
                    if (parameter.type.isMarkedNullable) {
                        b.controlFlow("%L.let", parameterName) {
                            b.controlFlow("if (it == null)") {
                                for (idx in sqlParameter.sqlIndexes) {
                                    b.addCode(nativeType.bindNull("_stmt", idx + 1)).addCode("\n")
                                }
                                b.nextControlFlow("else")
                                for (idx in sqlParameter.sqlIndexes) {
                                    b.addCode(nativeType.bind("_stmt", "it", idx + 1)).addCode("\n")
                                }
                            }
                        }
                    } else {
                        for (idx in sqlParameter.sqlIndexes) {
                            b.addCode(nativeType.bind("_stmt", parameterName, idx + 1)).addCode("\n")
                        }
                    }
                } else {
                    for (idx in sqlParameter.sqlIndexes) {
                        b.addStatement("%N.set(_stmt, %L, %N)", parameterMapperName(function, parameter.variable), idx + 1, parameterName)
                    }
                }
            }
            if (parameter is QueryParameter.EntityParameter) {
                for (field in parameter.entity.fields) {
                    val fieldPropertyName = field.property.simpleName.getShortName()
                    val fieldName = "$parameterName?.$fieldPropertyName"
                    val sqlParameter = queryWithParameters.find(parameter.name + "." + fieldPropertyName)
                    if (sqlParameter == null || sqlParameter.sqlIndexes.isEmpty()) {
                        continue
                    }
                    val nativeType = JdbcNativeTypes.findNativeType(field.type.toTypeName())
                    val mapping = field.mapping.getMapping(JdbcTypes.jdbcParameterColumnMapper)
                    if (nativeType != null && mapping == null) {
                        if (parameter.type.isMarkedNullable || field.type.isMarkedNullable) {
                            b.controlFlow("%N?.%L.let", parameterName, fieldPropertyName) {
                                b.controlFlow("if (it == null)") {
                                    for (idx in sqlParameter.sqlIndexes) {
                                        b.addCode(nativeType.bindNull("_stmt", idx + 1)).addCode("\n")
                                    }
                                    b.nextControlFlow("else")
                                    for (idx in sqlParameter.sqlIndexes) {
                                        b.addCode(nativeType.bind("_stmt", "it", idx + 1)).addCode("\n")
                                    }
                                }
                            }
                        } else {
                            for (idx in sqlParameter.sqlIndexes) {
                                b.addCode(nativeType.bind("_stmt", "$parameterName.$fieldPropertyName", idx + 1)).addCode("\n")
                            }
                        }
                    } else {
                        for (idx in sqlParameter.sqlIndexes) {
                            b.addStatement("%N.set(_stmt, %L, %L)", parameterMapperName(function, parameter.variable, fieldPropertyName), idx + 1, fieldName)
                        }
                    }
                }
            }
        }
        if (batchParam != null) {
            b.addStatement("_stmt.addBatch()")
            b.endControlFlow()
        }
    }
}
